flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/fs/ForStFlinkFileSystem.java

```diff
@@ -0,0 +1,153 @@
+import java.net.URI;
+
+/**
+ * A {@link FileSystem} delegates some requests to file system loaded by Flink FileSystem mechanism.
+ *
+ * <p>All methods in this class maybe used by ForSt, please start a discussion firstly if it has to
+ * be modified.
+ */
+public class ForStFlinkFileSystem extends FileSystem {
+
+    private final FileSystem delegateFS;
+
+    public ForStFlinkFileSystem(FileSystem delegateFS) {
+        this.delegateFS = delegateFS;
+    }
+
+    /**
+     * Returns a reference to the {@link FileSystem} instance for accessing the file system
+     * identified by the given {@link URI}.
+     *
+     * @param uri the {@link URI} identifying the file system.
+     * @return a reference to the {@link FileSystem} instance for accessing the file system
+     *     identified by the given {@link URI}.
+     * @throws IOException thrown if a reference to the file system instance could not be obtained.
+     */
+    public static FileSystem get(URI uri) throws IOException {
+        return new ForStFlinkFileSystem(FileSystem.get(uri));
+    }
+
+    /**
+     * Create ByteBufferWritableFSDataOutputStream from specific path which supports to write data
+     * to ByteBuffer with {@link org.apache.flink.core.fs.FileSystem.WriteMode#OVERWRITE} mode.
+     *
+     * @param path The file path to write to.
+     * @return The stream to the new file at the target path.
+     * @throws IOException Thrown, if the stream could not be opened because of an I/O, or because a
+     *     file already exists at that path and the write mode indicates to not overwrite the file.
+     */
+    public ByteBufferWritableFSDataOutputStream create(Path path) throws IOException {
+        return create(path, WriteMode.OVERWRITE);
+    }
+
+    @Override
+    public ByteBufferWritableFSDataOutputStream create(Path path, WriteMode overwriteMode)
+            throws IOException {
+        return new ByteBufferWritableFSDataOutputStream(delegateFS.create(path, overwriteMode));
+    }
+
+    @Override
+    public ByteBufferReadableFSDataInputStream open(Path path, int bufferSize) throws IOException {
+        return new ByteBufferReadableFSDataInputStream(
+                delegateFS.open(path, bufferSize), () -> delegateFS.open(path, bufferSize), 32);
+    }
+
+    @Override
+    public ByteBufferReadableFSDataInputStream open(Path path) throws IOException {
+        // TODO: make inputStreamCapacity configurable
+        return new ByteBufferReadableFSDataInputStream(
+                delegateFS.open(path), () -> delegateFS.open(path), 32);
+    }
+
+    @Override
+    public boolean rename(Path src, Path dst) throws IOException {
+        // The rename is not atomic for ForSt. Some FileSystems e.g. HDFS, OSS does not allow a
+        // renaming if the target already exists. So, we delete the target before attempting the
+        // rename.
+        if (delegateFS.exists(dst)) {
+            boolean deleted = delegateFS.delete(dst, false);
```

Comment:
Does this carry the risk of accidental data deletion? In the case that the file exists, should we  thrown an exception instead of deleting the file?

Reply:
If I remember correctly, it is added because the usage of ForSt/RocksDB's `RenameFile` may overwrite the target file in some scenarios. e.g. when renaming `CURRENT` file.
Some FileSystems e.g. PosixFileSystem also just overwrite the target file when renaming.

Reply:
Thanks for your explanation, I got it.